### PR TITLE
Internalize IDataLoader

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Towards #https://github.com/dotnet/machinelearning/issues/1995

### CODE REVIEW

**Comment by TomFinley** [comment]
Thank you Ivan. What of the BinaryLoader? Should that remain public as well?

**Comment by TomFinley** [comment]
> Funny. I expect to get bunch of exceptions from any class which implements IDataLoader and public, but apparently that's not the case.
> 
> In reply to: [458766322](https://github.com/dotnet/machinelearning/pull/2309#issuecomment-458766322) [](ancestors = 458766322)

Right. So whether a method is public or not versus whether it adheres to the contract are, in some sense, independent. So you can have an internal interface, a public class can adhere to that interface with public methods. This FYI is why we often have the additional work, if the class must remain public, that its implementing methods become explicit, if that makes sense?

For now, it would be preferable if, as I think you have done, that the implementing classes *themselves* become internal, since we would want them to adhere not to the `IDataLoader` interface but its successor, the `IIDataReader` interface. (Though we may in the end wind up deleting `IDataLoader` and renaming what we now call `IDataReader`, `IDataLoader` in its place.)

File: src/Microsoft.ML.Data/DataLoadSave/PartitionedPathParser.cs [hunk 1]
```diff
@@ -193,12 +196,13 @@ public IEnumerable<string> ParseValues(string path)
 
     [TlcModule.Component(Name = ParquetPartitionedPathParser.LoadName, FriendlyName = ParquetPartitionedPathParser.UserName,
         Desc = ParquetPartitionedPathParser.Summary, Alias = ParquetPartitionedPathParser.ShortName)]
-    public class ParquetPartitionedPathParserFactory : IPartitionedPathParserFactory
+    internal class ParquetPartitionedPathParserFactory : IPartitionedPathParserFactory
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/PartitionedPathParser.cs]: >ParquetPartitionedPathParserFactory [](start = 19, length = 35)

Huh... weird. Not your code I know, just curious, is it obvious why this isn't living off in Parquet land?

File: src/Microsoft.ML.Data/Utilities/ComponentCreation.cs [hunk 2]
```diff
@@ -126,7 +126,8 @@ public static IDataView LoadTransforms(this IHostEnvironment env, Stream modelSt
         /// <summary>
         /// Creates a data loader from the arguments object.
         /// </summary>
-        public static IDataLoader CreateLoader<TArgs>(this IHostEnvironment env, TArgs arguments, IMultiStreamSource files)
+        [BestFriend]
+        internal static IDataLoader CreateLoader<TArgs>(this IHostEnvironment env, TArgs arguments, IMultiStreamSource files)
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/Utilities/ComponentCreation.cs]: >CreateLoader [](start = 36, length = 12)

Thanks Ivan. I wouldn't necessarily do this now, but this entire assembly seems like a *lot* of stuff we do not want public. But not as part of this PR.

File: src/Microsoft.ML.Data/DataLoadSave/PartitionedPathParser.cs (hunk 1 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/DataLoadSave/PartitionedPathParser.cs]: I don't know, and from your tone it looks like you want me to move them to Parquet project. Your wish is my command master.

---
In reply to: [252374825](https://github.com/dotnet/machinelearning/pull/2309#discussion_r252374825) [](ancestors = 252374825)

**Comment by artidoro** [comment]
>     /// we could simply replace this with an array of IFileHandle.

Could you move this Review comment as it is a public summary?

---
Refers to: src/Microsoft.ML.Data/Data/IDataLoader.cs:14 in 66de22e. [](commit_id = 66de22e97f4c2f0862386d3907e46ba85d01c76d, deletion_comment = False)

**Comment by artidoro** [comment]
>         /// </summary>

Here as well.

---
Refers to: src/Microsoft.ML.Data/Data/IDataLoader.cs:36 in 66de22e. [](commit_id = 66de22e97f4c2f0862386d3907e46ba85d01c76d, deletion_comment = False)

**Comment by artidoro** [comment]
>         public abstract class ArgumentsBase : TransformInputBase

Why not making `ArgumentsBase` and `Arguements` internal? As you did already in `ValueMapping`.

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueToKeyMappingTransformer.cs:111 in 66de22e. [](commit_id = 66de22e97f4c2f0862386d3907e46ba85d01c76d, deletion_comment = False)

**Comment by artidoro** [comment]
>         public abstract class ArgumentsBase : TransformInputBase

But we will make it internal anyways as part of the issue that cleans up the transformers.

---
In reply to: [459065845](https://github.com/dotnet/machinelearning/pull/2309#issuecomment-459065845) [](ancestors = 459065845)

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueToKeyMappingTransformer.cs:111 in 66de22e. [](commit_id = 66de22e97f4c2f0862386d3907e46ba85d01c76d, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         public abstract class ArgumentsBase : TransformInputBase

Yeah, I would prefer to not diverge from intent of this PR

---
In reply to: [459066115](https://github.com/dotnet/machinelearning/pull/2309#issuecomment-459066115) [](ancestors = 459066115,459065845)

---
Refers to: src/Microsoft.ML.Data/Transforms/ValueToKeyMappingTransformer.cs:111 in 66de22e. [](commit_id = 66de22e97f4c2f0862386d3907e46ba85d01c76d, deletion_comment = False)

File: src/Microsoft.ML.Parquet/PartitionedPathParser.cs [hunk 3]
```diff
@@ -276,8 +280,7 @@ public void Save(ModelSaveContext ctx)
                 ctx.SaveString(sb.ToString());
             };
         }
-
-        public IEnumerable<PartitionedFileLoader.Column> ParseColumns(string path)
+        IEnumerable<PartitionedFileLoader.Column> IPartitionedPathParser.ParseColumns(string path)
```
Comment (artidoro) [inline src/Microsoft.ML.Parquet/PartitionedPathParser.cs]: >IEnumerable<PartitionedFileLoader.Column> IPartitionedPathParser.ParseColumns(string path) [](start = 8, length = 90)

Why this change if the outer class is internal? Also you didn't change `ParseColumns` above, which makes me thing it does not need to be changed.

File: src/Microsoft.ML.Data/Transforms/ValueMapping.cs [hunk 4]
```diff
@@ -362,7 +362,7 @@ internal bool TryUnparse(StringBuilder sb)
             }
         }
 
-        public sealed class Arguments
+        internal sealed class Arguments
```
Comment (artidoro) [inline src/Microsoft.ML.Data/Transforms/ValueMapping.cs]: >internal [](start = 8, length = 8)

For consistency, I would either make this public and only make specific field internal, or make the `Arguments` internal in the other transforms too.

File: src/Microsoft.ML.Parquet/PartitionedPathParser.cs (hunk 3 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Parquet/PartitionedPathParser.cs]: I made this change first, and only after that made whole class internal. Thanks for the catch!

---
In reply to: [252397204](https://github.com/dotnet/machinelearning/pull/2309#discussion_r252397204) [](ancestors = 252397204)

File: src/Microsoft.ML.Data/Transforms/ValueMapping.cs (hunk 4 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/Transforms/ValueMapping.cs]: In other places we use `Arguments` classes as EntryPoint arguments. I don't want to get into business of cleaning entrypoints in this PR.
For this transform we don't have entrypoint, so I prefer to just hide it completely. 

---
In reply to: [252398379](https://github.com/dotnet/machinelearning/pull/2309#discussion_r252398379) [](ancestors = 252398379)

Comment (artidoro) [reply src/Microsoft.ML.Data/Transforms/ValueMapping.cs]: Ok sounds good!

---
In reply to: [252399795](https://github.com/dotnet/machinelearning/pull/2309#discussion_r252399795) [](ancestors = 252399795,252398379)

File: src/Microsoft.ML.Data/DataLoadSave/PartitionedPathParser.cs (hunk 1 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/DataLoadSave/PartitionedPathParser.cs]: Well, that would certainly be very nice, thank you @Ivanidzo4ka. 😄

File: src/Microsoft.ML.Data/Utilities/ComponentCreation.cs (hunk 2 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/Utilities/ComponentCreation.cs]: Oh what I meant was the whole class. But like I said not part of this PR.
